File: 0113-path-sum-ii/0113-path-sum-ii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number[][]}
 */
var pathSum = function (root, targetSum) {

    function solve(root, targetSum, arr, res, sum) {
        if (root == null) {
            return;
        }
        if (root.left==null && root.right==null &&  root.val + sum == targetSum) {
            const arrcopy = [...arr, root.val];
            res.push(arrcopy)
        }
        solve(root.left, targetSum, [...arr, root.val], res, sum + root.val)
        solve(root.right, targetSum, [...arr, root.val], res, sum + root.val)
    }
    if (root == null) {
        return []
    }

    const res = []
    const arr = []
    var sum = 0;

    solve(root, targetSum, arr, res, sum);

    return res;





};

File: 0113-path-sum-ii/0113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    void dfs(TreeNode* root, vector<int>& path, vector<vector<int>>& result,
             int sum, int targetSum) {
        if (root == NULL)
            return;
        path.push_back(root->val);
        if (root->left==NULL && root->right==NULL && sum + root->val == targetSum) {
            result.push_back(path);
        }

        dfs(root->left, path, result, sum + root->val, targetSum);
        dfs(root->right, path, result, sum + root->val, targetSum);
        path.pop_back();
    }
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {

        int sum = 0;

        vector<int> path;

        vector<vector<int>> result;

        dfs(root, path, result, sum, targetSum);

        return result;
    }
};